### ISSUE TITLE
weight extraction

### ISSUE THREAD

**Issue by Sameeksha0709** [larq/larq#910]
hi, i am extracting the weights but weights are in floating point. I am using conversion if weights are negative taking it as -1 and if positive taking it as +1. But accuracy is dropping with these. First test accuracy was 92.94% and then it dropped to 28.1%

### Environment

TensorFlow version: 
Larq version:

**Comment by CNugteren** [comment]
Something like this should do the job:
```python
with larq.context.quantized_scope(True):
    weights = model.layers[1].get_weights()  # example layer '1'
    print(weights[0])  # binarized weights
    print(weights[1])  # floating-point bias terms
```

As for running inference with efficiency thanks to the binary weights you got with Larq, see https://github.com/larq/compute-engine/.

**Comment by Sameeksha0709** [comment]
Hi with this code also i am getting floating values of weight

**Comment by CNugteren** [comment]
Can you share a bit more of what you're doing?

**Comment by Sameeksha0709** [comment]
```python
kwargs = dict(input_quantizer="ste_sign",
              kernel_quantizer="ste_sign",
              kernel_constraint="weight_clip")

model = tf.keras.models.Sequential()
model.add(lq.layers.QuantConv2D(4, (5, 5),
                                kernel_quantizer="ste_sign",
                                kernel_constraint="weight_clip",
                                use_bias=False,strides = (2,2),activation='relu',
                                input_shape=(28, 28, 1)))
model.add(tf.keras.layers.BatchNormalization(scale=False))
model.add(tf.keras.layers.Flatten())

model.add(lq.layers.QuantDense(128, use_bias=False, **kwargs))
model.add(tf.keras.layers.BatchNormalization(scale=False))
model.add(lq.layers.QuantDense(10, use_bias=False, **kwargs))

model.add(tf.keras.layers.Activation("softmax"))
lq.models.summary(model)
```

Hi, this is my model. I want to extract the weights weights i need in integer type  because i am using the model on edge devices and for inference i have to write my own code without using any library

**Comment by Sameeksha0709** [comment]
i want binary weights.

**Comment by CNugteren** [comment]
Are you sure you didn't make a mistake with running the code? If I take your code, add some imports, and then add the following I only get 1's and -1's:
```python
with larq.context.quantized_scope(True):
    weights = model.layers[0].get_weights()  # example layer '0'
    print(weights[0])  # binarized weights
```

**Comment by Sameeksha0709** [comment]
Hi, I want to work in completely integer domain but batch normalization weights are in floating. Is there any way for that i can get integer value. If i am removing batch normalization then accuracy is dropping

**Comment by CNugteren** [comment]
You can have a look at LCE (https://github.com/larq/compute-engine/), which uses models trained in Larq and doesn't use floating-point. Basically batch normalization can be fused within the bias value of the preceding layer.

By the way, see [this recent issue](https://github.com/larq/larq/issues/907) in which a model is made to be fully integer.

**Comment by Sameeksha0709** [comment]
Hi, i am using following code 

```python
(train_images, train_labels), (test_images, test_labels) = tf.keras.datasets.mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
test_images = test_images.reshape((10000, 28, 28, 1))

# Scale pixel values to be between 0 and 4 and convert to integers
train_images = ((train_images / 255.0) * 4).astype(int)
test_images = ((test_images / 255.0) * 4).astype(int) 
kwargs = dict(input_quantizer="ste_sign",
              kernel_quantizer="ste_sign",
              kernel_constraint="weight_clip")

model = tf.keras.models.Sequential()
model.add(lq.layers.QuantConv2D(4, (5, 5),
                                kernel_quantizer="ste_sign",
                                kernel_constraint="weight_clip",
                                use_bias=False,strides = (2,2),activation='relu',
                                input_shape=(28, 28, 1)))
model.add(tf.keras.layers.BatchNormalization(scale=False))
model.add(tf.keras.layers.Flatten())

model.add(lq.layers.QuantDense(128, use_bias=False, **kwargs))
model.add(tf.keras.layers.BatchNormalization(scale=False))
model.add(lq.layers.QuantDense(10, use_bias=False, **kwargs))

model.add(tf.keras.layers.Activation("softmax"))
lq.models.summary(model)
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(train_images, train_labels, batch_size=64, epochs=10)
test_loss, test_acc = model.evaluate(test_images, test_labels)
print(f"Test accuracy {test_acc * 100:.2f} %")

model = tf.keras.models.load_model("BNN_model.h5")
converted = lce.convert_keras_model(model)
with open("my_model.tflite", "wb") as f:
    f.write(converted) 
```

when i am h5  to convert to tflite i am getting 
![my_model tflite](https://github.com/user-attachments/assets/f0f60df1-702b-4a10-8526-103ae931cffd)
in between i am getting add and multiply and middle layer weights are in float 
and with another conversion 

```python
import tensorflow as tf
from larq_compute_engine import convert_keras_model

# Path to save the TFLite model
tflite_model_path = '/home/user/Documents/sam/BNN_model.tflite'

# Load the trained model
model_path = '/home/user/Documents/sam/BNN_model'
model = tf.keras.models.load_model(model_path)

# Convert to TensorFlow Lite
with open(tflite_model_path, "wb") as flatbuffer_file:
    flatbuffer_bytes = convert_keras_model(
        model,
        inference_input_type=tf.int8,  # Specify input as 8-bit integers
        inference_output_type=tf.float32,  # Specify output as 32-bit floats
        experimental_default_int8_range=(0, 4)  # Default quantization range
    )
    flatbuffer_file.write(flatbuffer_bytes)

print(f"Model successfully converted to TFLite and saved at {tflite_model_path}") 
```

![BNN_model tflite](https://github.com/user-attachments/assets/e04ba9bd-6c19-4cf0-be71-293ded2eb256)
i am also getting the same. 
My requirement is that i want input in integer value and all weights in integers prefereably binary and for inference without using any library i have to write the functions for convolution operation. so i need only weights.

**Comment by CNugteren** [comment]
If you don't want to use the Larq Compute Engine (LCE) for inference, then I suggest you don't convert the model to TFLite. The easiest way to get the binary weight values out from the model is what I said above in https://github.com/larq/larq/issues/910#issuecomment-2551290755. Here is the full script based on your example code: 
```python
import tensorflow as tf
import larq as lq

kwargs = dict(input_quantizer="ste_sign",
              kernel_quantizer="ste_sign",
              kernel_constraint="weight_clip")

model = tf.keras.models.Sequential()
model.add(lq.layers.QuantConv2D(4, (5, 5),
                                kernel_quantizer="ste_sign",
                                kernel_constraint="weight_clip",
                                use_bias=False,strides = (2,2),activation='relu',
                                input_shape=(28, 28, 1)))
model.add(tf.keras.layers.BatchNormalization(scale=False))
model.add(tf.keras.layers.Flatten())

model.add(lq.layers.QuantDense(128, use_bias=False, **kwargs))
model.add(tf.keras.layers.BatchNormalization(scale=False))
model.add(lq.layers.QuantDense(10, use_bias=False, **kwargs))

model.add(tf.keras.layers.Activation("softmax"))
lq.models.summary(model)

with lq.context.quantized_scope(True):
    for layer_id in (0, 3, 5):
        weights = model.layers[layer_id].get_weights()
        print(weights[0])  # binarized weights
```

**Comment by Sameeksha0709** [comment]
Hi, but in inference I have to add batch normalization weights which is in floating value. Without batch normalization accuracy is dropping.

**Comment by Sameeksha0709** [comment]
model = tf.keras.models.Sequential()
model.add(lq.layers.QuantConv2D(4, (5, 5),
                                kernel_quantizer="ste_sign",
                                kernel_constraint="weight_clip",
                                kernel_initializer="glorot_uniform",
                                #input_quantizer= "ste_sign",
                                padding="same",
                                pad_values=1,
                                use_bias=False,strides = (2,2),activation='relu',
                                input_shape=(28, 28, 1)))
model.add(tf.keras.layers.BatchNormalization())
model.add(tf.keras.layers.Flatten())
model.add(lq.layers.QuantDense(10, use_bias=False, kernel_quantizer="ste_sign",kernel_constraint="weight_clip"))
model.add(tf.keras.layers.Activation("softmax"))
lq.models.summary(model)
i have simplify the model. I need only weights as binary basically BWN. and input as integer is also ok. i am using mnist dataset for this which has value either 0 or 1. only problem is that Batch normalization weights which are in float. how to fix that

**Comment by CNugteren** [comment]
We can't give support for writing your own inference engine, you'll have to figure out yourself how to do that. You can likely fuse those computations into the bias values of your preceding layers, but I can't help you with the details here. That is beyond the scope of Larq. There is likely plenty of information about this topic on the internet.

**Comment by Sameeksha0709** [comment]
import tensorflow as tf
from larq_compute_engine import convert_keras_model

# Path to save the TFLite model
tflite_model_path = '/home/user/Documents/sam/BNN_model.tflite'

# Load the trained model
model_path = '/home/user/Documents/sam/BNN_model'
model = tf.keras.models.load_model(model_path)

# Convert to TensorFlow Lite
with open(tflite_model_path, "wb") as flatbuffer_file:
    flatbuffer_bytes = convert_keras_model(
        model,
        inference_input_type=tf.int8,  # Specify input as 8-bit integers
        inference_output_type=tf.float32,  # Specify output as 32-bit floats
        experimental_default_int8_range=(-1, 1)  # Default quantization range
    )
    flatbuffer_file.write(flatbuffer_bytes)

print(f"Model successfully converted to TFLite and saved at {tflite_model_path}")

after using this i am getting mul and add values in int which can be used in place of batch normalization. but weights are coming values of -128 and 127. is there any way so i can weight -1 and 1
